nixos/modules/services/x11/redshift.nix

```diff
@@ -93,52 +34,59 @@ in {
         redshift derivation to use.
       '';
     };
-
-    extraOptions = mkOption {
-      type = types.listOf types.str;
-      default = [];
-      example = [ "-v" "-m randr" ];
-      description = ''
-        Additional command-line arguments to pass to
-        <command>redshift</command>.
-      '';
-    };
   };
 
   config = mkIf cfg.enable {
-    assertions = [ 
-      {
-        assertion = 
-          if cfg.provider == "manual"
-          then (cfg.latitude != null && cfg.longitude != null) 
-          else (cfg.latitude == null && cfg.longitude == null);
-        message = "Latitude and longitude must be provided together, and with provider set to null.";
-      }
-    ];
+    assertions =
+      with lib; let
+        allThe = x: !(any (map isNull x));
+        no = x: all (map isNull x);
+        oneOf = xs: (count id xs) == 1;
+        together = x: {
+          assertion = (allThe x.items) || (no x.items);
+          message = x.names + " must be provided together.";
+        };
+        coordinates = [ cfg.config.manual.lat cfg.config.manual.lon ];
+        times = [ cfg.config.redshift.dawn-time cfg.config.redshift.dusk-time ];
+        mechanisms = [ (allThe coordinates) (allThe times) ];
+        manualLocationProvider = cfg.config.redshift.location-provider == "manual";
+      in
+        (map together [ 
+          {
+            names = "Latitude and longitude";
+            items = coordinates;
+          }
+          {
+            names = "Time of dawn and time of dusk";
+            items = times;
+          }
+        ])
+        ++
+        [
+          {
+            assertion = oneOf mechanisms;
+            message = "The transitions must be set via exactly one mechanism.";
```

Comment:
What happens if both are set? This assertion seems a bit unnecessary

Reply:
Nothing bad, the manually set times will take precedence; but I thought if we had a chance to only allow configurations that follow the user's expectations and disallow ambiguous ones where we could that'd be a win.